WIP: Default span constraint on index

takojunior opened the pull request:
The PR is for adding a default index attribute and a default maximum span constraint on this attribute. This is going to avoid regular users to run into scaling issues when they have long sequences in data and usually apply no constraints at the beginning.

The PR has the following changes:

- Given a new parameter `max_span_index` to initialize Seq2Pat, add the default constraint. `max_span_index=10` by default.
- A fix to backend for checking upper and lower bounds in Gap constraints, where we may want to use original gap value ( a_i - a_{i-1} ) instead of the abs value | a_i - a_{i-1} |. With this change, I can get expected results when gap has negative upper/lower bounds in the tests, where it works together with default span cs. For this change, I might want to ask aminhn to confirm that it is a safe change.
- Update unit tests to align with the default constraint. All tests get passed.

I like to have the above changes reviewed first, then notebooks and docs can be updated with the default constraint.

skadio commented on sequential/backend/build_mdd.cpp:
```diff
@@ -98,16 +98,18 @@ bool Check_gap(int i, int strp, int endp, vector<vector<vector<int> > >* attrs,
-		if ((*lgap)[att] == 0)
-			continue;
-		if (abs((*attrs)[(*lgapi)[att]].at(i).at(endp - 1) - (*attrs)[(*lgapi)[att]].at(i).at(strp - 1)) < (*lgap)[att])
+// Allow lower bound of gap constraint to be 0 if not work with abs values
```
This is the fix that Amin did in the original/earlier repository, correct? 

My suggestions is; go that repo and look at the latest commits. It should be easy to find the commit for this fix at the backend. Then, you can compare & contrast the solutions. 

I vaguely remember looking into it at the time (~2 years ago) and it was a simple/one-liner change like this. 

It should be easy to verify.

takojunior replied:
I see his original repo uses abs value at this line though (see the [lines](https://github.com/aminhn/MPP/blob/master/src/build_mdd.cpp#L81)), which is the latest commit. 

In an earlier commit, there is a fix at another line ([line 46](https://github.com/aminhn/MPP/commit/009157d2a547b5481aa03bce169f755b3c2d3f4a#diff-b6d92d609a9a4838760571e76d844b5908d71648ecd54860becb1979c595037bR46)), which was to apply abs instead of removing it. So I tried to do the same fix as Amin did, which didn't give me expected results as I have negative upper/lower bounds and run it on paper though.

I am thinking that maybe I can leave an inline comment (shown as an issue) to the original repo and Amin can provide his thoughts there?

skadio commented on sequential/seq2pat.py:
```diff
@@ -277,11 +277,17 @@ class Seq2Pat:
 
-    def __init__(self, sequences: List[list]):
-        # Validate input sequences
+    def __init__(self, sequences: List[list], max_span_index: Union[int, None] = 10):
```
Not sure about the naming.. The "index" is about the Attribute but this parameter is about the "length" of the span. It is not index, really. 

So I would suggest `max_span` instead. 

On the contrary, `index_attr = Attribute()` in the code below makes total sense --because that thing is about the index

takojunior replied:
`max_span` makes sense. I will change the naming.

skadio reviewed (approved):
LGTM! This is a small change but will make the behavior much better for new comers / first experience with the library. Thank you!

skadio reviewed (approved):
In the usage example notebook, you might want to add a comment about this new parameter. 

Btw, shall we consider changing that notebook name to "sequential_pattern_mining" to complement the other notebook "dichomotic_pattern_mining"? 

And once this change is merged notebooks and pypi updated etc. there are a few threads we can circle back: 
 
* The current github issue might be solved in the next version
* XAI notebook you shared can be updated, Shovon et al should install new version
* Danielle should install new version

takojunior commented:
> In the usage example notebook, you might want to add a comment about this new parameter.
> 
> Btw, shall we consider changing that notebook name to "sequential_pattern_mining" to complement the other notebook "dichomotic_pattern_mining"?
> 
> And once this change is merged notebooks and pypi updated etc. there are a few threads we can circle back:
> 
> * The current github issue might be solved in the next version
> * XAI notebook you shared can be updated, Shovon et al should install new version
> * Danielle should install new version

With this update, I will circle back with these threads. 

For the dichotomic pattern mining note, it can be simplified now by dropping the span constraint and having the default constraint being in use.

takojunior commented:
The latest commit has the following updates to this PR:

- Refactored the parameter for built-in span constraint to be `max_span` in Seq2Pat
- Refactored the `rolling_window_size` parameter to be `max_span` for consistency in Pat2Feat
- Updated `usage_example` notebook to `sequential_pattern_mining` notebook. Added a comment to introduce `max_span` and the built-in constraint
- Updated `dichotomic pattern mining` notebook by dropping the manually created attribute and then utilized the built-in constraint. 
- Updated docs in the API pages.

@skadio @dorukkilitcioglu Let me know if you have further comments. Thanks.

skadio commented on sequential/pat2feat.py:
```diff
@@ -15,18 +15,18 @@ class _OneHotEncoding:
     """
 
-    def __init__(self, rolling_window_size: Union[int, None] = 10):
+    def __init__(self, max_span: Union[int, None] = 10):
```
ah good point on making this consistent!

skadio reviewed (approved):
LGTM --thank you @takojunior! This will very much simplify and speed-up everything else (notebooks, examples, etc.)

dorukkilitcioglu commented on sequential/seq2pat.py:
```diff
@@ -277,11 +277,18 @@ class Seq2Pat:
     sequences : List[list]
         A list of sequences each with a list of events.
         The event values can be all strings or all integers.
+    max_span: Union[int, None]
```
Any reason why `Union[int, None]` instead of `Optional[int]`? It accomplishes the same thing either way, but I'm wondering if you had a preference.

dorukkilitcioglu reviewed (approved):
These changes look good! Left a couple of comments, but good to go as it is.

takojunior commented on notebooks/dichotomic_pattern_mining.ipynb:
```diff
@@ -265,9 +251,9 @@
     "## Seq2Pat for Positive Labels\n",
-    "- There are two attributes: `event_time` and `event_order`\n",
-    "- Constraint 1: to enforce the average event time greater than 20 sec\n",
-    "- Constraint 2: to enforce the span of event order less than 9. This is to restrict the length of sequence to be <= 10."
```
Yes, the span constraint is now a built-in constraint and optional to use.

takojunior commented on sequential/backend/build_mdd.cpp:
```diff
@@ -98,16 +98,18 @@ bool Check_gap(int i, int strp, int endp, vector<vector<vector<int> > >* attrs,
 
 	for (int att = 0; att < (*lgap).size(); att++){
-		if ((*lgap)[att] == 0)
+// It enables the case when lower bound of gap constraint is 0. If lower bound is not a number, continue
```
Will add a note to point out the differences.
